Indicators/AverageDirectionalIndex.cs

```diff
@@ -0,0 +1,332 @@
+        private decimal ComputeTrueRange(TradeBar input)
+        {
+            var trueRange = new decimal(0.0);
+
+            if (_previousInput == null) return trueRange;
+
+            trueRange = (Math.Max(Math.Abs(input.Low - _previousInput.Close), Math.Max(TrueRange.Current, Math.Abs(input.High - _previousInput.Close))));
+
+            return trueRange;
+        }
+
+        /// <summary>
+        /// Computes the positive directional movement.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <returns></returns>
+        private decimal ComputePositiveDirectionalMovement(TradeBar input)
+        {
+            var postiveDirectionalMovement = new decimal(0.0);
+
+            if (_previousInput == null) return postiveDirectionalMovement;
+
+            if ((input.High - _previousInput.High) >= (_previousInput.Low - input.Low))
+            {
+                if ((input.High - _previousInput.High) > 0)
+                {
+                    postiveDirectionalMovement = input.High - _previousInput.High;
+                }
+            }
+
+            return postiveDirectionalMovement;
+        }
+
+        /// <summary>
+        /// Computes the negative directional movement.
+        /// </summary>
+        /// <param name="input">The input.</param>
+        /// <returns></returns>
+        private decimal ComputeNegativeDirectionalMovement(TradeBar input)
+        {
+            var negativeDirectionalMovement = new decimal(0.0);
+
+            if (_previousInput == null) return negativeDirectionalMovement;
+
+            if ((_previousInput.Low - input.Low) > (input.High - _previousInput.High))
+            {
+                if ((_previousInput.Low - input.Low) > 0)
+                {
+                    negativeDirectionalMovement = _previousInput.Low - input.Low;
+                }
+            }
+
+            return negativeDirectionalMovement;
+        }
+
+        /// <summary>
+        /// Computes the next value of this indicator from the given state
+        /// </summary>
+        /// <param name="input">The input given to the indicator</param>
+        /// <returns>A new value for this indicator</returns>
+        protected override decimal ComputeNextValue(TradeBar input)
+        {
+            TrueRange.Update(input);
+            DirectionalMovementPlus.Update(input);
+            DirectionalMovementMinus.Update(input);
+            SmoothedTrueRange.Update(Current);
+            SmoothedDirectionalMovementMinus.Update(Current);
+            SmoothedDirectionalMovementPlus.Update(Current);
+            PositiveDirectionalIndex.Update(Current);
+            NegativeDirectionalIndex.Update(Current);
+            var diff = Math.Abs(PositiveDirectionalIndex - NegativeDirectionalIndex);
+            var sum = PositiveDirectionalIndex + NegativeDirectionalIndex;
+            var value = Math.Round((sum == 0 ? 50 : ((_period - 1) * Current.Value + 100 * diff / sum) / _period), 5);
```

Comment:
Is rounding a part of the definition? I typically like to leave these unrounded, allows the caller to decide rounding behavior.